win32/tests/file_win32_ut/file_win32_ut.c

```diff
@@ -0,0 +1,1011 @@
+/*Tests_SRS_FILE_WIN32_43_002: [ file_create shall call SetFileCompletionNotificationModes to disable calling the completion port when an async operations finishes synchrounously. ]*/
+/*Tests_SRS_FILE_WIN32_43_003: [ file_create shall initialize a threadpool environment by calling InitializeThreadpolEnvironment. ]*/
+/*Tests_SRS_FILE_WIN32_43_004: [ file_create shall obtain a PTP_POOL struct by calling execution_engine_win32_get_threadpool on execution_engine. ]*/
+/*Tests_SRS_FILE_WIN32_43_005: [ file_create shall register the threadpool environment by calling SetThreadpoolCallbackPool on the initialized threadpool environment and the obtained ptp_pool ]*/
+/*Tests_SRS_FILE_WIN32_43_006: [ file_create shall create a cleanup group by calling CreateThreadpoolCleanupGroup. ]*/
+/*Tests_SRS_FILE_WIN32_43_007: [ file_create shall register the cleanup group with the threadpool environment by calling SetThreadpoolCallbackCleanupGroup. ]*/
+/*Tests_SRS_FILE_WIN32_43_033: [ file_create shall create a threadpool io with the allocated FILE_HANDLE and on_file_io_complete_win32 as a callback by calling CreateThreadpoolIo ]*/
+/*Tests_SRS_FILE_WIN32_43_009: [ file_create shall succeed and return a non-NULL value. ]*/
+TEST_FUNCTION(file_create_succeeds)
+{
+    ///arrange
+    PTP_CALLBACK_ENVIRON captured_ptpcbe;
+    PTP_WIN32_IO_CALLBACK captured_callback = NULL;
+    char* filename = "file_create_succeeds.txt";
+    setup_file_create_expectations(filename, &captured_ptpcbe, &captured_callback);
+
+    ///act
+    FILE_HANDLE file_handle = file_create(fake_execution_engine, filename, NULL, NULL);
+
+    ///assert
+    ASSERT_IS_NOT_NULL(file_handle);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    
+    ///cleanup
+    file_destroy(file_handle);
+}
+
+/*Tests_SRS_FILE_43_034: [ If there are any failures, file_create shall fail and return NULL. ]*/
+/*Tests_SRS_FILE_WIN32_43_008: [ If there are any failures, file_create shall return NULL. ]*/
+TEST_FUNCTION(file_create_fails)
+{
+    ///arrange
+    const char filename[] = "file_create_succeeds.txt";
+    PTP_CALLBACK_ENVIRON captured_ptpcbe;
+    PTP_WIN32_IO_CALLBACK captured_ptp_callback;
+
+    setup_file_create_expectations(filename, &captured_ptpcbe, &captured_ptp_callback);
+
+    umock_c_negative_tests_snapshot();
+    for (int i = 0; i < umock_c_negative_tests_call_count(); ++i)
+    {
+        if (umock_c_negative_tests_can_call_fail(i))
+        {
+            umock_c_negative_tests_reset();
+            umock_c_negative_tests_fail_call(i);
+
+            ///act
+            FILE_HANDLE file_handle = file_create(fake_execution_engine, filename, NULL, NULL);
+
+            ///assert
+            ASSERT_IS_NULL(file_handle);
+
+            ///cleanup
+            file_destroy(file_handle);

```

Comment:
not needed, create returned NULL

Reply:
I put it there if file_create doesn't fail. I thought that we should still cleanup even if the test fails.

---
In reply to: [457772798](https://github.com/Azure/azure-c-pal/pull/10#discussion_r457772798) [](ancestors = 457772798)

Reply:
If the test fails, we bail out and this code isn't executed

---
In reply to: [458134864](https://github.com/Azure/azure-c-pal/pull/10#discussion_r458134864) [](ancestors = 458134864,457772798)